### PR TITLE
[FEA] ARM support for Modulus Container

### LINKED ISSUE #132: 🚀[FEA]: ARM support for Modulus Docker container
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem you would like to solve.

Modulus docker container is currently only supported for x86_64 and amd64 architectures. Would be nice to introduce support for arm64 architectures. 

### Describe any alternatives you have considered

_No response_

### PR DESCRIPTION
**Pull request by ktangsali**
# Modulus Pull Request

## Description
This PR adds support for docker build on ARM architecture. To build a multi-arch image, the below command can be used:

```
docker buildx build --platform linux/amd64,linux/arm64 -t modulus:deploy-multiarch --target deploy -f Dockerfile .
```

Closes #132 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -15,6 +15,8 @@
 ARG PYT_VER=23.06
 FROM nvcr.io/nvidia/pytorch:$PYT_VER-py3 as builder
 
+ARG TARGETPLATFORM
```
Comment (mnabian) [inline Dockerfile:18]: Should we default this to `linux/amd64` to simplify our predominant build command?

File: pyproject.toml [hunk 2]
```diff
@@ -13,7 +13,6 @@ requires-python = ">=3.8"
 license = {text = "Apache 2.0"}
 dependencies = [
     "h5py>=3.7.0",
-    "mpi4py>=3.1.4",
```
Comment (mnabian) [inline pyproject.toml:16]: I think we shouldn't silently remove these. At minimum we should add a proper import error message.

File: Dockerfile (hunk 1 above)
Comment (ktangsali) [reply Dockerfile:18]: I have tried this. The issue is, the default arg will work with `docker build` command. But with `docker buildx`, the default specified in the Dockerfile overrides the arg passed via the `--platform` command line arg. Thus, `docker buildx` will fail for say `arm64`.

**Review by mnabian — APPROVED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -29,38 +31,71 @@ ENV _CUDA_COMPAT_TIMEOUT=90
...
     fi
+
 # cleanup of stage
 RUN rm -rf /modulus/ 
 
 # CI image
 FROM builder as ci
-RUN pip install "tensorflow>=2.9.0" "warp-lang>=0.6.0" "black==22.10.0" "interrogate==1.5.0" "coverage==6.5.0" "protobuf==3.20.0" 
+
+ARG TARGETPLATFORM
```
Comment (NickGeneva) [inline Dockerfile:87]: Why is this here vs above on line 3?

File: Dockerfile [hunk 4]
```diff
@@ -73,7 +108,16 @@ RUN rm -rf /modulus/
 
 # Docs image
 FROM deploy as docs
+
+ARG TARGETPLATFORM
+
 # Install CI packages
-RUN pip install "tensorflow>=2.9.0" "warp-lang>=0.6.0" "protobuf==3.20.0"
+RUN pip install "protobuf==3.20.0"
```
Comment (NickGeneva) [inline Dockerfile:115]: curious why was protobufs added for docs?

File: Makefile [hunk 5]
```diff
@@ -46,11 +46,11 @@ coverage:
 		coverage html
 
 container-deploy:
-	docker build -t modulus:deploy --target deploy -f Dockerfile .
+	docker build -t modulus:deploy --build-arg TARGETPLATFORM=linux/amd64 --target deploy -f Dockerfile .
```
Comment (NickGeneva) [inline Makefile:49]: so how do you build with buildx? is that documented somewhere? maybe the release repo

Comment (NickGeneva) [reply Makefile:49]: I'm not familiar with the buildx builder and what not. Just some info for knowledge transfer somewhere even if its internal would be good to capture.

File: README.md [hunk 6]
```diff
@@ -95,18 +95,23 @@ pip install .
...
 ```bash
-docker build -t modulus:ci --target ci -f Dockerfile .
+docker build -t modulus:ci \
+    --build-arg TARGETPLATFORM=linux/amd64 --target ci -f Dockerfile .
 ```
 
-Alternatively, you can run `make container-ci`
+Alternatively, you can run `make container-ci`.
+
+Currently only `linux/amd64` and `linux/arm64` platforms are supported. `linux/amd64`
```
Comment (NickGeneva) [inline README.md:113]: So for ARM, we just dont add a bunch of dependencies? That seems to be the case with the Dockerfile. Should there be a quick comment that some utils wont work in the ARM container?

File: Dockerfile (hunk 4 above)
Comment (NickGeneva) [reply Dockerfile:115]: Nevermind this is in CI image section!

File: Makefile (hunk 5 above)
Comment (ktangsali) [reply Makefile:49]: Yes, I had captured that in the PR description :). But will also document it in the release repo. The regular builds will work just fine for single arch image.

File: README.md (hunk 6 above)
Comment (ktangsali) [reply README.md:113]: No, all dependencies are supported with our build in ARM too. This Dockerfile will build the dependencies from source that don't have an released ARM wheel/package.

File: Dockerfile (hunk 3 above)
Comment (ktangsali) [reply Dockerfile:87]: Yeah, unfortunately these variables don't propagate across stages.

File: Makefile (hunk 5 above)
Comment (akshaysubr) [inline Makefile:49]: Would be good to set `TARGETPLATFORM` based on the detected CPU arch on the system. Something like

```make
ARCH := $(shell uname -p)

ifeq ($(ARCH), x86_64)
    TARGETPLATFORM := "linux/amd64"
else
    ifeq ($(ARCH), aarch64)
        TARGETPLATFORM := "linux/arm64"
    else
        $(error Unknown CPU architecture ${ARCH} detected)
    endif
endif

container-deploy:
	docker build -t modulus:deploy --build-arg TARGETPLATFORM=${TARGETPLATFORM} --target deploy -f Dockerfile .
```

File: Dockerfile [hunk 7]
```diff
@@ -29,38 +31,71 @@ ENV _CUDA_COMPAT_TIMEOUT=90
...
 # TODO use torch-harmonics pip package after the upgrade
 RUN pip install https://github.com/NVIDIA/torch-harmonics/archive/8826246cacf6c37b600cdd63fde210815ba238fd.tar.gz
-RUN pip install "tensorly>=0.8.1" "vtk>=9.2.6" "pyvista>=0.40.1" https://github.com/tensorly/torch/archive/715a0daa7ae0cbdb443d06780a785ae223108903.tar.gz
+RUN pip install "tensorly>=0.8.1" https://github.com/tensorly/torch/archive/715a0daa7ae0cbdb443d06780a785ae223108903.tar.gz
+
+# copy modulus source
+COPY . /modulus/
+
+# install vtk and pyvista
+RUN if [ "$TARGETPLATFORM" = "linux/arm64" ]; then \
```
Comment (akshaysubr) [inline Dockerfile:40]: Why is this `arm64` instead of `aarch64`? Similarly for `amd64` instead of `x86_64`? Would be nice to align this to the output of `uname -p` if there is no other constraint

File: README.md [hunk 8]
```diff
@@ -95,18 +95,23 @@ pip install .
...
+is the recommended default.
```
Comment (akshaysubr) [inline README.md:114]: Maybe leave out the recommended default? It's not like there is a choice to make once you already have your system.

File: CHANGELOG.md [hunk 9]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Added distributed utilities to create process groups and orthogonal process groups.
 - Added distributed AFNO model implementation.
+- Add instructions for docker build on ARM architecture.
```
Comment (akshaysubr) [inline CHANGELOG.md:15]: Add -> Added

File: Dockerfile [hunk 10]
```diff
@@ -29,38 +31,71 @@ ENV _CUDA_COMPAT_TIMEOUT=90
...
+RUN if [ "$TARGETPLATFORM" = "linux/amd64" ] && [ -e "/modulus/deps/dgl-1.1.1-cp310-cp310-linux_x86_64.whl" ]; then \
+        echo "DGL wheel for $TARGETPLATFORM exists, installing!" && \
+        pip install --force-reinstall /modulus/deps/dgl-1.1.1-cp310-cp310-linux_x86_64.whl; \
+    elif [ "$TARGETPLATFORM" = "linux/arm64" ] && [ -e "/modulus/deps/dgl-1.1.1-cp310-cp310-linux_aarch64.whl" ]; then \
+        echo "DGL wheel for $TARGETPLATFORM exists, installing!" && \
+        pip install --force-reinstall /modulus/deps/dgl-1.1.1-cp310-cp310-linux_aarch64.whl; \
+    else \
+        echo "No DGL wheel present, building from source" && \
+	git clone https://github.com/dmlc/dgl.git && cd dgl/ && git checkout tags/1.1.1 && git submodule update --init --recursive && \
 	DGL_HOME="/workspace/dgl" bash script/build_dgl.sh -g && \
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply Dockerfile:61]: Yeah, the DGL source build just works :)

File: Dockerfile [hunk 11]
```diff
@@ -29,38 +31,71 @@ ENV _CUDA_COMPAT_TIMEOUT=90
...
+ARG TARGETPLATFORM
+
+RUN pip install "black==22.10.0" "interrogate==1.5.0" "coverage==6.5.0" "protobuf==3.20.0" "mip4py>=3.1.4"
 COPY . /modulus/
 RUN cd /modulus/ && pip install -e . && rm -rf /modulus/
+RUN if [ "$TARGETPLATFORM" = "linux/amd64" ]; then \
+	echo "Installing tensorflow and warp-lang for: $TARGETPLATFORM" && \
+	pip install "tensorflow>=2.9.0" "warp-lang>=0.6.0"; \ 
+    elif [ "$TARGETPLATFORM" = "linux/arm64" ]; then \
+	echo "Installing tensorflow and warp-lang for: $TARGETPLATFORM is not supported presently"; \
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply Dockerfile:96]: I think warp team is looking into it. Tensorflow should have worked but for some reason it fails in the docker build (works fine if installed manually inside a built container). Since tensorflow would be removed soon, and our CI targets only use x86_64 arch or amd64 arch, I thought it should be fine to exclude tensorflow at the moment. I can look more into it

File: Makefile (hunk 5 above)
Comment (ktangsali) [reply Makefile:49]: This is great snippet! Will add this

File: Dockerfile (hunk 7 above)
Comment (ktangsali) [reply Dockerfile:40]: I think they are normalized anyways, hence I was using the normalized values directly (reference: https://github.com/containerd/containerd/blob/v1.4.3/platforms/platforms.go#L86) and their official docs used the above names. Do we need to change in that case (because it will normalize anyways)?

File: README.md (hunk 8 above)
Comment (ktangsali) [reply README.md:114]: Good point haha!

File: Dockerfile (hunk 10 above)
[the remark this answers is not in the document]
Comment (akshaysubr) [reply Dockerfile:61]: Have we tested running something that uses DGL on ARM? Even if it builds, it might've built for x86_64, but will throw a runtime error on aarch64

File: Dockerfile (hunk 7 above)
Comment (akshaysubr) [reply Dockerfile:40]: If there is a precedent for defining it the way we have it, then let's leave it as is. Maybe just add a documentation item pointing to where we got this convention from.

File: Dockerfile (hunk 10 above)
[the remark this answers is not in the document]
Comment (ktangsali) [reply Dockerfile:61]: Yes, I have tested pytests for meshgraphnet and graphcast and it worked on an ARM machine.

File: README.md (hunk 8 above)
Comment (ktangsali) [reply README.md]: Done! Also added a note about `warp-lang` support on arm64

File: Makefile (hunk 5 above)
Comment (ktangsali) [reply Makefile]: Added

File: Dockerfile (hunk 7 above)
Comment (ktangsali) [reply Dockerfile:40]: Added a line in the Makefile

File: Dockerfile (hunk 11 above)
[the remark this answers is not in the document]
Comment (ktangsali) [reply Dockerfile:96]: Added a note to the README. Let me know if that would suffice.

**Review by NickGeneva — APPROVED** [review]

**Review by akshaysubr — APPROVED** [review]
